### PR TITLE
[Merged by Bors] - use atx grading for hare

### LINKED ISSUE #4089: implement hare active set
## background

prior to unified content block, the hare active set was derived from contextually valid blocks verified by tortoise, hence are in consensus.
after unified content block, the hare active set was derived from first ballot of epoch from smeshers. since nodes do not have consensus on ballots, this active set is not in consensus

## proposal

- implement https://community.spacemesh.io/t/grading-atxs-for-the-active-set/335
- hare active set for epoch E will be derived from the first block in epoch E in the following way:
  - take the proposals produced by the ATXIDs recorded in the block
  - the union of the active set referenced by said proposals is used as the hare active set for epoch E (plus safe layer range)
- newly synced node (not online last epoch to receive ATXs real time) will use the first block of current epoch and for the active set in its ballot

## bootstrapping and recovery
a spacemesh signed active set (a list of ATX IDs) will b used for
- the first safe layer range in epoch 2.
- when hare fails to reach consensus after N layers for any given epoch

## implementation plan
- [ ] https://github.com/spacemeshos/go-spacemesh/pull/4148 use atx grading scheme in reference ballot and hare
- [x] https://github.com/spacemeshos/go-spacemesh/pull/4231 binary that publishes bootstrap and fallback values for beacon and active set
- [x] https://github.com/spacemeshos/go-spacemesh/pull/4241 bootstrap and use first consensus block for hare active set

### LINKED ISSUE #4757: add test for ActiveSetFromBlock
## Description
following up on https://github.com/spacemeshos/go-spacemesh/pull/4717

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation

Closes #4089
Closes #4757

## Changes
- hare only vote for proposals with grade 1/acceptable and 2/good atxs in the activeset

### CODE REVIEW

File: hare/hare.go [hunk 1]
```diff
@@ -441,9 +442,21 @@ func (h *Hare) removeCP(ctx context.Context, lid types.LayerID) {
...
 
 // goodProposals finds the "good proposals" for the specified layer. a proposal is good if
-// it has the same beacon value as the node's beacon value.
+// - it has the same beacon value as the node's beacon value.
+// - its miner is not malicious
+// - its active set contains only grade 1 or grade 2 atxs
+// see (https://community.spacemesh.io/t/grading-atxs-for-the-active-set/335#proposal-voting-4)
 // any error encountered will be ignored and an empty set is returned.
-func goodProposals(ctx context.Context, logger log.Log, msh mesh, nodeID types.NodeID, lid types.LayerID, epochBeacon types.Beacon) []types.ProposalID {
+func goodProposals(
```
Comment (fasmat) [inline hare/hare.go:450]: It seems that besides the changes in this file only a few tests were cleaned up compared to https://github.com/spacemeshos/go-spacemesh/pull/4717.

If this is the case the PR LGTM

**Review by fasmat — APPROVED** [review]

File: miner/atx_grader.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+// grade 2: ATX was received at time t < s-4δ, and no equivocation proof was received for that id until time s.
+type AtxGrade int
+
+const (
+	Evil AtxGrade = iota
+	Acceptable
+	Good
+)
+
+func GradeAtx(msh mesh, nodeID types.NodeID, atxReceived, epochStart time.Time, delta time.Duration) (AtxGrade, error) {
```
Comment (dshulyak) [inline miner/atx_grader.go]: this function can be more generic to make it reusable, i think there is no need for it to load malfeasance proof.
there could be one more on top that loads it and can error

`func GradeAtx(atxReceived, proofReceived, epochStart time.Time, delta time.Duration) Grade`

File: miner/atx_grader.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+// let s be the start of the epoch, and δ the network propagation time.
+// grade 0: ATX was received at time t >= s-3δ, or an equivocation proof was received by time s-δ.
+// grade 1: ATX was received at time t < s-3δ before the start of the epoch, and no equivocation proof was received by time s-δ.
+// grade 2: ATX was received at time t < s-4δ, and no equivocation proof was received for that id until time s.
+type AtxGrade int
+
+const (
+	Evil AtxGrade = iota
+	Acceptable
+	Good
```
Comment (dshulyak) [inline miner/atx_grader.go]: can we add two more grades? it also makes sense to call them Grade1-5 explicitly, as this Evil/Acceptable/Good won't work anymore

i don't remember deltas in those grades, but there should be a picture somewhere

**Review by dshulyak — APPROVED** [review]
i realized that we don't persist grades, so it doesn't matter that much if this change is 3-graded or 5-graded

File: miner/util.go [hunk 4]
```diff
@@ -12,7 +12,7 @@ import (
 	"github.com/spacemeshos/go-spacemesh/sql/certificates"
 )
 
-func ActiveSetFromBlock(db sql.Executor, epoch types.EpochID) ([]types.ATXID, error) {
+func ActiveSetFromEpochFirstBlock(db sql.Executor, epoch types.EpochID) ([]types.ATXID, error) {
```
Comment (fasmat) [inline miner/util.go:15]: ❤️

File: miner/util_test.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/datastore"
+	"github.com/spacemeshos/go-spacemesh/log/logtest"
+	"github.com/spacemeshos/go-spacemesh/signing"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/blocks"
+	"github.com/spacemeshos/go-spacemesh/sql/certificates"
+)
+
+func TestActiveSetFromEpochFirstBlock(t *testing.T) {
```
Comment (fasmat) [inline miner/util_test.go:18]: Thanks for adding the test ❤️

**Review by fasmat — APPROVED** [review]

**Comment by countvonzero** [comment]
bors merge

**Pull request closed without merging**